content/blog/2020-04-preemption.md

```diff
@@ -0,0 +1,183 @@
+is assigned a budget of 128 operations. When `socket.read(..)` and
+`socket.write(..)` are called, the budget is decremented. If the budget is zero,
+the task yields back to the scheduler. If either `read` or `write` cannot
+proceed due to the underlying socket not being ready (no pending data or a full
+send buffer), then the task also yield back to the scheduler.
+
+The idea originated from a conversation I was having with [Ryan Dahl][ry]. He is
+using Tokio as the underlying runtime for [Deno][deno]. When doing some HTTP
+experimentation (a while back) with [Hyper], he was seeing some high tail
+latencies in some benchmarks. The problem was due to a loop not yielding back to
+the scheduler under load. Hyper ended up fixing the problem by hand in this one
+case, but Ryan mentioned that, when he worked on [node.js][node], they handled
+the problem by adding **per resource** limits. So, if a TCP socket was always
+ready, it would force a yield every so often. I mentioned this conversation to
+[Jon Gjenset][jonhoo], and he ended up coming up with the idea of placing the
+limit on the task itself instead of on each resource.
+
+The end result is that Tokio should be able to provide more consistent runtime
+behavior under load. While the exact heuristics will most likely be tweaked over
+time, initial measurements show that, in some cases, tail latencies are reduced
+almost 3x.
+
+[![benchmark](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)](https://user-images.githubusercontent.com/176295/73222456-4a103300-4131-11ea-9131-4e437ecb9a04.png)
+
+"master" is before the automatic yielding and "preempt" is after. Click for a
+bigger version, also see the original [PR comment][pr] for more details.
+
+## A note on blocking
+
+While automatic cooperative task yielding improves many cases, it cannot preempt
+tasks. Users of Tokio must still take care to avoid CPU intensive work and using
+blocking APIs. The [`spawn_blocking`][spawn_blocking] function can be used to
+"asyncify" these sorts of tasks by running them on a thread pool.
+
+Tokio does not, and will not attempt to detect blocking tasks and automatically
+compensate by adding threads to the scheduler. This question has come up a
+number of times in the past so allow me to elaborate.
+
+For context, the idea is for the scheduler to include a monitoring thread. This
+thread would poll scheduler threads every so often and check that workers are
+making progress. If a worker is not making progress, it is assumed that the
+worker is executing a blocking task and a new thread should be spawned to
+compensate.
+
+This idea is not new. The first occurence of this strategy that I am aware of is
+in the .NET thread pool and was introduced more than ten years ago.
+Unfortunately, the strategy has a number of problems and because of this, has
+not been featured in other thread pools / schedulers (golang, java, erlang,
```

Comment:
```suggestion
Unfortunately, the strategy has a number of problems and because of this, it has
not been featured in other thread pools / schedulers (golang, java, erlang,
```